### ISSUE TITLE
QueryEntity selectMany using auditTime(0)

### ISSUE THREAD

**Issue by FERNman** [datorama/akita#286]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Current behavior

QueryEntity ```selectMany``` returns an Obseravle with ```auditTime(0)```, which appends the return event to the JS event loop. Plain ```select``` calls, on the other hand, return immediately, causing a delay between the two.

## Expected behavior

```selectMany``` should return immediately, just like ```select```.

## Minimal reproduction of the problem with instructions

https://stackblitz.com/edit/github-xbhyct
When you navigate to the TODO page, ```selectMany``` and ```selectCount``` are called. Although both could (and should) return at the same time, ```selectMany``` returns one change cycle later (see the alert message).

## What is the motivation / use case for changing the behavior?

We are currently implementing virtual scrolling and are sending an HTTP request to get the first x entities and the total entity count. Both of these are then plugged into our component, which calls a method to load more entities as soon as the user scrolls near the bottom of the table and as long as the total entity count is bigger than the fetched entity count. When the total count gets set one change cycle earlier, the user is obviously near the end of the table (as there are no entities, but the count is already loaded) and the 'load more' lambda is triggered.

In general, I'd say this can lead to confusion as it's unexpected.

## Environment

<pre><code>
Angular version: 8.0.6

Browser:
- [x] Chrome (desktop) version 76
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 10.16.0  
- Platform:  Windows 

Others:

</code></pre>

**Comment by NetanelBasal** [comment]
This is by design. Otherwise, the observable will fire per entity which doesn't want we want. We want to group them to one `next`.

**Comment by FERNman** [comment]
Ah, I see... Well, that's unfortunate. Thanks for the reply though!

**Comment by FERNman** [comment]
Ok so I played around with it a little bit more and I think something like the following could do the job:

```typescript
selectMany<R>(ids: IDType[]): Observable<EntityType[]>;
selectMany<R>(ids: IDType[], project: (entity: EntityType) => R): Observable<R[]>;
selectMany<R>(ids: IDType[], project?: (entity: EntityType) => R): Observable<EntityType[] | R[]> {
  if (!ids || !ids.length) return of([]);

  return this.select(state => state.entities).pipe(
    map(entities => ids.map(id => getEntity(id, project)(entities))),
    map((entities): R[] | EntityType[] => entities.filter(Boolean)),
    distinctUntilArrayItemChanged()
  ) as Observable<R[] | EntityType[]>;
}
```

The casts are only necessary to keep the signatures the same, as `distinctUntilArrayItemChanged()`  returns `<(R | EntityType)[]>`. I think this isn't as fast as the previous approach, but it also shouldn't be much of a performance killer.
All the unit tests are green, except `Many > SelectMany > should select many`, because this method returns immediately instead of waiting in the queue. Therefore, the callback gets invoked twice.

**Comment by NetanelBasal** [comment]
That's nice.

> I think this isn't as fast as the previous approach, but it also shouldn't be much of a performance killer.

Could you do a benchmark with a large store, please?

**Comment by FERNman** [comment]
I did a bit of benchmarking, and, at least for my tests, the new code is actually a bit faster than the old one.

I ran 25 rounds of tests for each of the two test cases. Each time, a new store was created with 5000 ToDo Entries and the first 1000 of them were queried using selectMany.
Afterwards, I updated the first 200 entries in the selected range. I updated 100 one-by-one by id, then I updated the following 100 all at once.
I also updated the 200 entries not in the selected range, again 100 one by one and 100 at once.

With the new code, I got around 1700ms for these functions, the old one took about 2000ms on average. I hope this tests the code as intended!
